File: Pyramid Transition Matrix/Pyramid Transition Matrix.py
# Solution for Pyramid Transition Matrix in PY

from typing import List, Dict

class Solution:
    def pyramidTransition(self, bottom: str, allowed: List[str]) -> bool:

        # Build transition map: bottom pair -> possible top blocks
        transitions: Dict[str, List[str]] = {}
        for pattern in allowed:
            transitions.setdefault(pattern[:2], []).append(pattern[2])

        memo = {}

        def build_pyramid(current_row: str, index: int, next_row: str) -> bool:
            # Base case: reached the top
            if len(current_row) == 1:
                return True

            # Memoization key
            state_key = (current_row, index, next_row)
            if state_key in memo:
                return memo[state_key]

            # Finished processing current row → move up
            if index == len(current_row) - 1:
                memo[state_key] = build_pyramid(next_row, 0, "")
                return memo[state_key]

            # Current adjacent block pair
            block_pair = current_row[index:index + 2]

            # No valid transition
            if block_pair not in transitions:
                memo[state_key] = False
                return False

            # Try all possible top blocks
            for top_block in transitions[block_pair]:
                if build_pyramid(current_row, index + 1, next_row + top_block):
                    memo[state_key] = True
                    return True

            memo[state_key] = False
            return False

        # Start DFS from bottom row
        return build_pyramid(bottom, 0, "")

File: Pyramid Transition Matrix/Pyramid Transition Matrix.java
// Solution for Pyramid Transition Matrix in JAVA

class Solution {

    /**
     * Recursive helper to determine if a valid pyramid can be built.
     *
     * @param currentRow  Current row being processed
     * @param transitions Map of allowed bottom-pair → possible top blocks
     * @param index       Current index in currentRow
     * @param nextRow     Row being built above currentRow
     * @param memo        Memoization cache
     */
    private boolean buildPyramid(String currentRow,
                                 Map<String, List<Character>> transitions,
                                 int index,
                                 StringBuilder nextRow,
                                 Map<String, Boolean> memo) {

        // Base case: reached the top of the pyramid
        if (currentRow.length() == 1) {
            return true;
        }

        // Memoization key captures full recursive state
        String stateKey = currentRow + "_" + index + "_" + nextRow.toString();
        if (memo.containsKey(stateKey)) {
            return memo.get(stateKey);
        }

        // Finished processing current row → move to next row
        if (index == currentRow.length() - 1) {
            boolean result = buildPyramid(
                nextRow.toString(), transitions, 0, new StringBuilder(), memo
            );
            memo.put(stateKey, result);
            return result;
        }

        // Current adjacent block pair
        String blockPair = currentRow.substring(index, index + 2);

        // No valid transition
        if (!transitions.containsKey(blockPair)) {
            memo.put(stateKey, false);
            return false;
        }

        // Try all possible top blocks
        for (char topBlock : transitions.get(blockPair)) {
            nextRow.append(topBlock);

            if (buildPyramid(currentRow, transitions, index + 1, nextRow, memo)) {
                memo.put(stateKey, true);
                return true;
            }

            // Backtrack
            nextRow.deleteCharAt(nextRow.length() - 1);
        }

        memo.put(stateKey, false);
        return false;
    }

    public boolean pyramidTransition(String bottom, List<String> allowed) {

        // Build transition map
        Map<String, List<Character>> transitions = new HashMap<>();
        for (String pattern : allowed) {
            transitions
                .computeIfAbsent(pattern.substring(0, 2), k -> new ArrayList<>())
                .add(pattern.charAt(2));
        }

        // Memoization cache
        Map<String, Boolean> memo = new HashMap<>();

        // Start DFS from bottom row
        return buildPyramid(bottom, transitions, 0, new StringBuilder(), memo);
    }
}
